### PR TITLE
Session modal on agenda page

### PR DESCRIPTION
**Pull request by weepotty**
This PR styles the session modal on the agenda page.

A few issues I ran into when styling this modal:
1. On loading the modal, I wanted to be able to disable the scrolling of the page 'behind' the modal when the modal pops up. Used javascript + `overflow: hidden` for this. Also included a blurring/darkening of the background on modal load - is there a way to do this without JS?

2. I used turbo streams to append the modal close 'x' button.. never used turbo streams before, not sure if this is legit.

3. Couldn't work out how to navigate to the previous page using turbo - I read the docs on the part about Restoration visits, but 'Restoration visits have an action of restore and Turbo Drive reserves them for internal use. You should not attempt to annotate links or invoke Turbo.visit with an action of restore.' Is there some other data attribute that I can use? For now I have resorted to using `history.back()` on clicking the 'x'.

This is very much a draft, wanted to hear your thoughts before I do the speaker modal as well. Thanks! 😄

### CODE REVIEW

**Comment by ayushn21** [comment]
@weepotty this is really great ... you've got all the basics down. I picked up your work and just refactored it a bit. Here's the answers to your questions and some other reasoning behind my refactor:

I've refactored the JavaScript into a custom element. It keeps this nice and isolated and we'll need some more JS logic when making the modals accessible. Inline scripts are also generally a rather bad idea. In this project we have to use them in some cases because there's no JS build pipeline, but it's best to use them as a last resort when no other solution is viable.

Inline scripts also cause issues with Turbo Drive. That's why I've disabled Turbo for links in the nav bar and in breadcrumbs. Without that, the confetti button breaks. I don't understand why that happens well enough to explain it to you, but it's got something to do with they way Turbo Drive works.

> On loading the modal, I wanted to be able to disable the scrolling of the page 'behind' the modal when the modal pops up. Used javascript + overflow: hidden for this. Also included a blurring/darkening of the background on modal load - is there a way to do this without JS?

JS is the only way, good call here. I've moved the code into the custom element though.

> I used turbo streams to append the modal close 'x' button.. never used turbo streams before, not sure if this is legit.

This was actually a really good idea. I hadn't thought of this. In a web app you'd absolutely do it this way because you have some server side logic which would know when a modal is being presented. In a static site though, the cross was appearing even when it was presented as a full page making the use of a Turbo Stream unnecessary.

I've put it in just as an element now and hidden it using CSS unless it's inside a modal. Also, never use `div`s for interactive elements. That's really bad for accessibility as `div`s are not interactive by default. I've changed it to a button and put in an SVG for the cross.

> Couldn't work out how to navigate to the previous page using turbo - I read the docs on the part about Restoration visits, but 'Restoration visits have an action of restore and Turbo Drive reserves them for internal use. You should not attempt to annotate links or invoke Turbo.visit with an action of restore.' Is there some other data attribute that I can use? For now I have resorted to using history.back() on clicking the 'x'.

`history.back()` is perfect. It's the right way to do it. That's the great thing about Turbo is that everything is separate pages and then the transition is smooth-ed out using JavaScript. That's why to dismiss the modal, all we do is simulate a press of the browser's "back" button which is done using `history.back()`.


Let me know if you have any further questions. I'll try to work on making the modal accessible when I get some time, but once you're back, let's show a modal for the speaker cards in the carousel using the same technique.

**Comment by ayushn21** [comment]
@weepotty I've done some additional work on making the modals and agenda page accessible. Here's a gist of the things I've done:

- Add a `role=tabgroup` attribute to the links for each day. This helps screen readers to provide additional context.
- Jump focus to the first session item when a day is selected. This helps both screen readers and keyboard uses so the first session in that day is selected when a day is clicked.
- Dismiss a modal using the escape key.
- Make the background of the page "inert" when a modal is displayed. This ensured focus is trapped within the modal and the background is not selectable by keyboard and screen reader users.
- Return focus to previously active element when a modal is dismissed.

There's quite a bit going on there so let me know if you have any questions or if anything is confusing. Happy to get on a call to walk you through everything if you like.

**Comment by ayushn21** [comment]
@weepotty I just pushed a commit to make the breadcrumbs accessible: https://github.com/radioactivetoyhq/rubyonrails-website/pull/52/commits/e5556837134fd5f1cfccde15e36eaa26ff93e3b9.

Quick note on the implementation there, I missed it first time round. The immediate descendent of a `ul` or an `ol` has to be an `li`. We can't have other tags as direct descendants. That's why I've moved the `>` symbols into the `li` tags. Could you please replicate the changes across to the other breadcrumbs as well?

I know I've done quite a lot on this PR. It's ballooned a bit. Try to go through it commit by commit. I reckon you'll have a few questions around what I've done, so drop me a message once you're back and we can chat about the best way to proceed! Might be worth getting on a call to talk everything through before merging this.

**Comment by weepotty** [comment]
This is awesome! Feel like there's lots here for me to learn. Would be great to hop on a call to chat through it. Will ping you on Basecamp. I've just pushed a commit to fix the speaker breadcrumb.

File: _layouts/world/speaker.html [hunk 1]
```diff
@@ -5,11 +5,19 @@
 
 <content-container class="speaker">
     <breadcrumbs-container data-turbo="false">
```
Comment (ayushn21) [inline _layouts/world/speaker.html:7]: Let's add `aria-label="Breadcrumbs"` as an attribute here ... it helps with accessibility for screen reader users.

**Pull request merged**